lib/real-device-management.js

```diff
@@ -72,15 +73,71 @@ async function installToRealDevice(device, app, bundleId, opts) {
     return;
   }
 
-  const {skipUninstall, strategy, timeout} = opts ?? {};
+  const {skipUninstall, strategy, timeout, shouldEnforceUninstall} = opts ?? {};
+  const shouldUninstallApp = !skipUninstall && bundleId;
 
-  if (!skipUninstall && bundleId && (await device.isAppInstalled(bundleId))) {
+  if (shouldUninstallApp) {
     log.debug(`Reset requested. Removing app with id '${bundleId}' from the device`);
     await device.remove(bundleId);
   }
   log.debug(`Installing '${app}' on device with UUID '${device.udid}'...`);
-  await device.install(app, timeout, strategy);
-  log.debug('The app has been installed successfully.');
+
+  try {
+    await device.install(app, timeout, strategy);
+    log.debug('The app has been installed successfully.');
+  } catch (e) {
+    // Do nothing if the session did not require to enforce uninstalling the application
+    if (!shouldEnforceUninstall) {
+      throw e;
+    }
+
+    if (!bundleId) {
+      log.warn(`An error occurred during installing '${app}'. No bundle id was given, ` +
+        `thus Appium have not tried to uninstall the app. Original error: ${e.message}`);
+      throw e;
+    }
+
+    // Want to clarify the device's application installation state in this situation.
+
+    // If the error was by below error case, we could recover the situation
+    // by uninstalling the device's app bundle id explicitly regard less the app exists on the device or not (e.g. offload app).
+    // [XCUITest] Error installing app '/path/to.app': Unexpected data: {"Error":"MismatchedApplicationIdentifierEntitlement","ErrorDescription":"Upgrade's application-identifier entitlement string (TEAM_ID.com.kazucocoa.example) does not match installed application's application-identifier string (ANOTHER_TEAM_ID.com.kazucocoa.example); rejecting upgrade."}
+    if (!shouldUninstallApp && e.message.includes('MismatchedApplicationIdentifierEntitlement')) {
+      log.info(`MismatchedApplicationIdentifierEntitlement error occurred because '${bundleId}' might remain as ` +
```

Comment:
How about:
```
The application identified by '{bundleId}' cannot be installed because it might be already cached on the device, probably with a different signature. Will try to remove it and install a new copy. Original error: ${e.message}'
```